components/modules/eth.c

```diff
@@ -0,0 +1,312 @@
+  //vTaskDelay(1);
+
+  if (enable)
+    module_config.eth_config->phy_power_enable(true);
+}
+
+/**
+ * @brief gpio specific init
+ *
+ * @note RMII data pins are fixed in esp32:
+ * TXD0 <=> GPIO19
+ * TXD1 <=> GPIO22
+ * TX_EN <=> GPIO21
+ * RXD0 <=> GPIO25
+ * RXD1 <=> GPIO26
+ * CLK <=> GPIO0
+ *
+ */
+static void eth_gpio_config_rmii( void )
+{
+  phy_rmii_configure_data_interface_pins();
+  phy_rmii_smi_configure_pins( module_config.pin_mdc, module_config.pin_mdio );
+}
+
+
+
+// --- Event handling -----------------------------------------------------
+
+typedef void (*fill_cb_arg_fn) (lua_State *L, const system_event_t *evt);
+typedef struct
+{
+  const char *name;
+  system_event_id_t event_id;
+  fill_cb_arg_fn fill_cb_arg;
+} event_desc_t;
+
+#define ARRAY_LEN(a) (sizeof(a) / sizeof(a[0]))
+
+static void eth_got_ip (lua_State *L, const system_event_t *evt);
+static void empty_arg (lua_State *L, const system_event_t *evt) {}
+
+static const event_desc_t events[] =
+{
+  { "start",            SYSTEM_EVENT_ETH_START,           empty_arg     },
+  { "stop",             SYSTEM_EVENT_ETH_STOP,            empty_arg     },
+  { "connected",        SYSTEM_EVENT_ETH_CONNECTED,       empty_arg     },
+  { "disconnected",     SYSTEM_EVENT_ETH_DISCONNECTED,    empty_arg     },
+  { "got_ip",           SYSTEM_EVENT_ETH_GOT_IP,          eth_got_ip    },
+};
+
+#define ARRAY_LEN(a) (sizeof(a) / sizeof(a[0]))
+static int event_cb[ARRAY_LEN(events)];
+
+static int eth_event_idx_by_id( const event_desc_t *table, unsigned n, system_event_id_t id )
+{
+  for (unsigned i = 0; i < n; ++i)
+    if (table[i].event_id == id)
+      return i;
+  return -1;
+}
+
+static int eth_event_idx_by_name( const event_desc_t *table, unsigned n, const char *name )
+{
+  for (unsigned i = 0; i < n; ++i)
+    if (strcmp( table[i].name, name ) == 0)
+      return i;
+  return -1;
+}
+
+
+static void eth_got_ip( lua_State *L, const system_event_t *evt )
+{
+  (void)evt;
+  tcpip_adapter_ip_info_t ip_info;
+
+  memset(&ip_info, 0, sizeof(tcpip_adapter_ip_info_t));
+  if (tcpip_adapter_get_ip_info( ESP_IF_ETH, &ip_info ) != ESP_OK) {
+    luaL_error( L, "error from tcpip_adapter_get_ip_info!" );
+  }
+
+  char ipstr[IP_STR_SZ] = { 0 };
+  ip4str( ipstr, &ip_info.ip );
+  lua_pushstring( L, ipstr );
+  lua_setfield( L, -2, "ip" );
+
+  ip4str( ipstr, &ip_info.netmask );
+  lua_pushstring( L, ipstr );
+  lua_setfield( L, -2, "netmask" );
+
+  ip4str(ipstr, &ip_info.gw );
+  lua_pushstring( L, ipstr );
+  lua_setfield( L, -2, "gw" );
+}
+
+static void on_event( const system_event_t *evt )
+{
+  int idx = eth_event_idx_by_id( events, ARRAY_LEN(events), evt->event_id );
+  if (idx < 0 || event_cb[idx] == LUA_NOREF)
+    return;
+
+  lua_State *L = lua_getstate();
+  lua_rawgeti( L, LUA_REGISTRYINDEX, event_cb[idx] );
+  lua_pushstring( L, events[idx].name );
+  lua_createtable( L, 0, 5 );
+  events[idx].fill_cb_arg( L, evt );
+  lua_call( L, 2, 0 );
```

Comment:
Ok, changed that to `lua_pcall()`.

I still can't get my head around the decision when to use `lua_pcal`l vs. plain `lua_call`. Is there a generic rule for guidance?

Reply:
Quite understandable, because so far we haven't had a "one true way" for this. With an unprotected `lua_call()` everything dies if an error escapes it, and you get a cryptic panic message about something, somewhere, having died. With `lua_pcall()`, the error gets caught, but again you don't get any useful feedback, but at least your whole device didn't crash. All of this will get resolved once we get Terry's next big patch in, at which point all modules should simply use the new `luaN_call()` instead, which will capture errors, print a helpful stack trace, and also not crash the device. It should be the best of all worlds, really. Soon(tm).